helix-vcs/src/git.rs

```diff
@@ -61,24 +61,46 @@ impl Git {
     }
 }
 
-impl DiffProvider for Git {
-    fn get_diff_base(&self, file: &Path) -> Result<Vec<u8>> {
+impl VcsProvider for Git {
+    fn root_for_file(&self, file: &Path) -> Result<Box<dyn VcsRoot>> {
         debug_assert!(!file.exists() || file.is_file());
         debug_assert!(file.is_absolute());
-
-        // TODO cache repository lookup
```

Comment:
This should be addressed in a sperate PR. I prefer a different interface (this kind of disjoint dynamic dispatch based approach will not work once we add filewatchjng to the mix). Repo caching is also sufficiently tricky and evolves resource based eviction and should just be it's ownPR (which also makes this one easier to review).

Reply:
oh, i didn't mean to actually remove this comment - the changes as they exist don't actually do any caching at all, they just always call `root_for_file` to recalculate the root (because yeah, doing the caching correctly would be tricky)

Reply:
Ah I see I just kind of skimmed it and it looked lime you were doing caching with the root behavior ut now I see that you still tried to keep the fallback behavior. I think we should just remove all.fallback behavior (and so t add this extra notion of a root)